torch/csrc/jit/python/pybind_utils.h

```diff
@@ -46,12 +46,20 @@
 namespace torch {
 namespace jit {
 
+py::object toPyObject(IValue ivalue);
+
 // The PythonFutureWrapper for ivalue::Future
-struct PythonFutureWrapper {
-  explicit PythonFutureWrapper(c10::intrusive_ptr<c10::ivalue::Future> fut)
-      : fut(std::move(fut)) {}
+struct VISIBILITY_HIDDEN PythonFutureWrapper {
+  using UnwrapFunc = std::function<void(py::object)>;
+
+  explicit PythonFutureWrapper(
+      c10::intrusive_ptr<c10::ivalue::Future> fut,
+      c10::optional<UnwrapFunc> unwrap_func = c10::nullopt)
+      : fut(std::move(fut)), unwrap_func(std::move(unwrap_func)) {}
 
-  IValue wait() {
+  PythonFutureWrapper(const PythonFutureWrapper&) = delete;
+
+  py::object wait() {
```

Comment:
Will need @wanchaol's comment on whether it is OK to change the return type of `PythonFutureWrapper::wait()`

Reply:
I was thinking, there is no need to split the code belong to a single API scattered into 2 files. `jit/python/init.cpp` and `jit/python.pybind_utils.h`, so I am merging them into a single place, and refer to this place using a function pointer `& PythonFutureWrapper::wait` in `jit/python/init.cpp`.

Reply:
I am ok with wait returning py::object here, but I don't think it's a good idea to pass a `unwrap` function here, it's a rpc future specific handling logic, we should just call it outside.

Reply:
@wanchaol 
I can't think of another way to correctly do this. 
I think adding a callback for whoever calls "wait()" is the way to go.
Now we only have callbacks for whoever calls markCompleted, setting the value on the future.

For client thread, all it has is the `jit::PythonFutureWrapper` and it will only call "wait()".
Server thread has access to the Future<Message> and knows it's an `RemoteException` object before everyone else. But as I mentioned in https://github.com/pytorch/pytorch/pull/35154/files#r396667781, server thread can't raise the exception.